src/lib/components/FileUpload/hooks.ts

```diff
@@ -0,0 +1,85 @@
+import { useCallback, useReducer } from "react";
+
+import { FileRejection } from "react-dropzone";
+
+import { FileUploadFile } from "./FileUpload";
+
+type FileUploadState = {
+  acceptedFiles: FileUploadFile[];
+  fileRejections: FileRejection[];
+};
+
+type AddFiles = {
+  type: "add-files";
+  payload: { acceptedFiles: FileUploadFile[]; fileRejections: FileRejection[] };
+};
+type RemoveAcceptedFile = { type: "remove-accepted"; payload: FileUploadFile };
+type RemoveRejectedFile = { type: "remove-rejected"; payload: FileRejection };
+
+type FileUploadActions =
+  | AddFiles
+  | RemoveAcceptedFile
+  | RemoveRejectedFile;
+
+export const useFileUpload = () => {
+  const fileUploadReducer = (
+    state: FileUploadState,
+    action: FileUploadActions,
+  ): FileUploadState => {
+    switch (action.type) {
+      case "add-files":
+        return {
+          acceptedFiles: [
+            ...state.acceptedFiles,
+            ...action.payload.acceptedFiles,
+          ],
+          fileRejections: [
+            ...state.fileRejections,
+            ...action.payload.fileRejections,
+          ],
+        };
+      case "remove-accepted": {
+        const newFiles = [...state.acceptedFiles];
+        newFiles.splice(newFiles.indexOf(action.payload), 1);
+        return {
+          acceptedFiles: [...newFiles],
+          fileRejections: [...state.fileRejections],
+        };
+      }
+      case "remove-rejected": {
+        const newRejectedFiles = [...state.fileRejections];
+        newRejectedFiles.splice(newRejectedFiles.indexOf(action.payload), 1);
+        return {
+          acceptedFiles: [...state.acceptedFiles],
+          fileRejections: [...newRejectedFiles],
+        };
+      }
```

Comment:
This could be made (arguably) easier to read and more concise by the use of filter.

```suggestion
      case "remove-rejected": {
        return {
          ...state,
          fileRejections: state.fileRejections.filter(
            (rejection) => rejection !== action.payload,
          ),
        };
      }
```